active/0011-jq-in-rule-engine.md

```diff
@@ -146,13 +147,67 @@ only output a single map result, and the map will then be piped to the JQ clause
 
 I am happy with this syntax now.
 
-### Introduce JQ as NIF
+### Introduce JQ as Port or NIF
 
+JQ is written in portable C as a single binary, reading command line argument
+from stdin and outputting results to the stdout. It supports Linux, OS X,
+FreeBSD, Solaris, and Windows for now, so the simplest way is to package the `jq`
+binary along with all of the emqx installation packages, and talk to `jq` using
+the [erlang port](http://erlang.org/doc/tutorial/c_port.html). For someone who
+is building emqx from source code of emqx repo, he can put the jq binary in to
+the right path according to the configuration.
 
+The second approach is NIF, with the drawback of more changes to the code (compile
+the code to a dynamic C library rather than a single binary), and safety (it
+brings down the entire erlang system on crash, and may hold up the erlang
+scheduler if it returns too late). But this way has the benefit of efficiency
+and no independent `jq` binary is required.
+
+### A try for using JQ as NIF
+
+An Erlang NIF library has been create at https://github.com/terry-xiaoyu/jqerl.
+
+It is now at the very early stage and is not well-tested.
+It requires the `jqlib`, so before using the library you should install the
```

Comment:
maybe it's in your plan already: compile jqlib source code ?

Reply:
Yes, let user install jq before building/installing emqx is not a good idea. It would lead to some problem if the use installed the wrong jq version.

I think we should also compile our `libcrypto`, as it is a common problem that emqx fails to start due to the wrong ssl version installed in the system.

Reply:
maybe more feasible to statically link crypto when building otp.
ref:  http://erlang.2086793.n4.nabble.com/static-build-of-Erlang-with-openssl-td4681642.html